### PR TITLE
"My Bids" module

### PR DESCRIPTION
**Pull request by ashfurrow**
https://github.com/artsy/auctions/issues/314

- [x] Initial render:
  - [x] We need to fetch the data to display these bids, if it's not already fetched.
  - [x] Throw together a ORStackView/UICollectionView (whichever is easiest) of bid cells. Going to try and re-use as much as possible from existing classes with the later goal of consolidating on one design.
  - [x] Insert that view into our existing `headerStack` view, under the banner.
  - [x] Connect tap gestures to new view controller pushes.
- [x] Updated render: 
  - [x] When the view re-appears, I think we should check for updated bids that have happened in the mean time.
- [x] Testing
  - [x] On a variety of devices
  - [x] Unit tests
  - [x] Apply design feedback
  - [x] Fix bug reported by Devang.
  - [ ] Fix Autolayout runtime warnings.

### CODE REVIEW

File: Artsy/Networking/API_Modules/ArtsyAPI+CurrentUserFunctions.m [hunk 1]
```diff
@@ -29,5 +30,16 @@ + (void)getCurrentUserBiddersForSale:(NSString *)saleID success:(void (^)(NSArra
...
 
++ (void)getCurrentUserLotStandingsForSale:(NSString *)saleID success:(void (^)(NSArray<SaleArtwork *> *))success failure:(void (^)(NSError *error))failure
+{
+    if ([[ARUserManager sharedManager] currentUser] == nil) {
+        success(@[]);
+        return;
+    }
+
+    NSURLRequest *request = [ARRouter lotStandingsRequestForSaleID:saleID];
+    [self getRequest:request parseIntoAnArrayOfClass:[SaleArtwork class] success:success failure:failure];
```
Comment (ashfurrow) [inline Artsy/Networking/API_Modules/ArtsyAPI+CurrentUserFunctions.m]: This is incorrect, [the API](https://github.com/artsy/gravity/blob/d364c85cff6fa8f2f9624023716e5f2f7857e1d1/app/services/lot_standing_service.rb#L2-L9) returns a sort of JSON pseudo-model with `bidder`, `sale_artwork`, `leading_position`, and `max_positions` fields.

**Comment by ArtsyOpenSource** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Error">
          1 Error
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:no_entry_sign:</td>
      <td data-sticky="false">CHANGELOG isn’t valid YAML</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          2 Warnings
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">PR is classed as Work in Progress</td>
    </tr>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">Needs testing on a Phone if change is non-trivial</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="http://danger.systems/">danger</a>
</p>

**Comment by ashfurrow** [comment]
Okay, title design is implemented. Now onto the hard stuff (notes follow).

simulator screen shot feb 13 2017 12 37 05 pm
simulator screen shot feb 13 2017 12 31 10 pm

Reusing the existing designs is going to be tricky, since the new ones are so different. We have an existing collection view cell that I think we can repurpose, from the list view of refine settings:

screen shot 2017-02-13 at 12 38 34 pm

Hopefully things are as easy as throwing a collection view into the new `LotStandingsView`, return a valid `intrinsicContentSize`, and have autolayout "just work". We'll see!

**Comment by ashfurrow** [comment]
Reusing existing `UICollectionView` infrastructure is a bust: its too tied to being inside the view controller hierarchy. Going to use an `ORStackView` with nibs instead, and we can reuse the nibs later inside collection views if that makes sense.

File: Artsy/Networking/ARRouter.m [hunk 2]
```diff
@@ -1153,6 +1153,17 @@ + (NSURLRequest *)biddersRequestForSale:(NSString *)saleID
     return [self requestWithMethod:@"GET" path:ARMyBiddersURL parameters:params];
 }
 
++ (NSURLRequest *)lotStandingsRequestForSaleID:(NSString *)saleID
+{
+    NSParameterAssert(saleID);
+    NSDictionary *params = @{
+        @"sale_id": saleID,
+        @"live": @(YES) // We only want to show active bids for open sales.
```
Comment (ashfurrow) [inline Artsy/Networking/ARRouter.m]: We decided against this, need to remove.

**Comment by ashfurrow** [comment]
Okay, got this to a good place:

simulator screen shot feb 14 2017 1 55 42 pm
simulator screen shot feb 14 2017 1 54 46 pm

@orta you'll be happy to hear that my experiences with xibs have been very positive. Even got to use the new stack view feature for the compact version (there're two, one inside another):

screen shot 2017-02-14 at 1 53 36 pm

It's pretty impressive how expressive you can be. Sadly I had to use regular layout constraints for the regular size class.

screen shot 2017-02-14 at 1 54 31 pm

File: Artsy/View_Controllers/Live_Auctions/Views/LotStandingsView.swift [hunk 3]
```diff
@@ -0,0 +1,57 @@
...
+        let bottomBorder = UIView().then {
+            $0.backgroundColor = UIColor.artsyGrayMedium()
+            $0.constrainHeight("1")
+        }
+        addSubview(bottomBorder)
+        bottomBorder.alignBottomEdge(withView: self, predicate: "0")
+        bottomBorder.alignLeading("0", trailing: "0", toView: self)
+
+        self.listView = listView
+        self.titleView = titleView
```
Comment (ashfurrow) [inline Artsy/View_Controllers/Live_Auctions/Views/LotStandingsView.swift]: Don't need these anymore.

File: Artsy/Views/Auction/LotStandingsLotListView.swift [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+private typealias PrivateFunctions = LotStandingsLotListView
+extension PrivateFunctions {
+    func setup() {
+        let numberOfLotStandings = saleViewModel.numberOfLotStandings
+
+        (0..<numberOfLotStandings)
+            .map { saleViewModel.lotStanding(at: $0) }
+            .enumerated()
+            .flatMap { (index, lotStanding) -> LotStandingsLotView? in
+                let drawBottomBorder = index != numberOfLotStandings - 1
```
Comment (ashfurrow) [inline Artsy/Views/Auction/LotStandingsLotListView.swift]: Wrap this in some parens.

**Comment by ashfurrow** [comment]
Okay, I've un-WIP'd this. Looks good, just want to test on devices instead of the sim. @orta handing this to you for review.

File: Artsy/View_Controllers/Auction/AuctionLotStandingsNetworkModel.swift [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+        ArtsyAPI.getCurrentUserLotStandings(
+            forSale: saleID,
+            success: { [weak self] lotStandings in
+                self?.lotStandings = lotStandings
+                observable.update(.success(lotStandings))
+            }, failure: { error in
+                observable.update(.error(error))
+        })
+
+        return observable
```
Comment (orta) [inline Artsy/View_Controllers/Auction/AuctionLotStandingsNetworkModel.swift]: 👍

File: Artsy/View_Controllers/Auction/AuctionNetworkModel.swift [hunk 6]
```diff
@@ -37,13 +39,19 @@ extension AuctionNetworkModel: AuctionNetworkModelType {
...
-        return fetchBidders()
-            .flatMap { [weak self] (bidders: Result<[Bidder]>) -> Observable<Result<SaleViewModel>> in
+        return fetchBidders().merge(lotStandingsNetworkModel.fetch(saleID))
+            .flatMap { [weak self] (bidders: Result<[Bidder]>, lotStandings: Result<[LotStanding]>) -> Observable<Result<SaleViewModel>> in
                 guard let `self` = self else { return Observable() }
 
+                var retrievedLotStandings = [LotStanding]()
+                if case .success(let lotStandings) = lotStandings {
+                    // Note: we're discarding any error in lotStandings because our UI doesn't expose errors.
+                    retrievedLotStandings = lotStandings
```
Comment (orta) [inline Artsy/View_Controllers/Auction/AuctionNetworkModel.swift]: good comment

File: Artsy/Views/Auction/LotStandingsLotViewCompact.xib [hunk 7]
```diff
@@ -0,0 +1,129 @@
+<?xml version="1.0" encoding="UTF-8"?>
+<document type="com.apple.InterfaceBuilder3.CocoaTouch.XIB" version="3.0" toolsVersion="11762" systemVersion="16D32" targetRuntime="iOS.CocoaTouch" propertyAccessControl="none" useAutolayout="YES" useTraitCollections="YES" colorMatched="YES">
+    <device id="retina4_7" orientation="portrait">
+        <adaptation id="fullscreen"/>
+    </device>
+    <dependencies>
+        <plugIn identifier="com.apple.InterfaceBuilder.IBCocoaTouchPlugin" version="11757"/>
+        <capability name="documents saved in the Xcode 8 format" minToolsVersion="8.0"/>
```
Comment (orta) [inline Artsy/Views/Auction/LotStandingsLotViewCompact.xib]: interesting, I never noticed this line in a xib

File: Artsy_Tests/View_Controller_Tests/Auction/AuctionLotStandingsNetworkModelSpec.swift [hunk 8]
```diff
@@ -0,0 +1,47 @@
...
+            expect(lotStandings).toNot( beEmpty() )
+        }
+
+
+        it("stores the sale artworks after fetching") {
+            waitUntil { done in
+                subject.fetch(saleID).subscribe { _ in
+                    done()
+                }
+            }
```
Comment (orta) [inline Artsy_Tests/View_Controller_Tests/Auction/AuctionLotStandingsNetworkModelSpec.swift]: are you sure you need these asyncs? `ArtsyAPI` is sync in testing

Comment (ashfurrow) [inline Artsy_Tests/View_Controller_Tests/Auction/AuctionLotStandingsNetworkModelSpec.swift]: Not sure, I'll test 👍

File: Artsy/Views/Auction/LotStandingsLotViewCompact.xib (hunk 7 above)
Comment (ashfurrow) [inline Artsy/Views/Auction/LotStandingsLotViewCompact.xib]: Might be new?

**Comment by ashfurrow** [comment]
Going to run some tests before merging. There are a few TODOs and cleanup items but this PR is already huge, and its shippable as-is, so I'll defer those to another PR.

**Comment by ashfurrow** [comment]
Design feedback:

background copy 2

**Comment by ashfurrow** [comment]
Reported by @devangt: 

image_0-3

Looks like the layout isn't re-flowing immediately. Might be related to the iPad issue Katarina noted in her design feedback.

**Comment by ashfurrow** [comment]
Just noticed some AL issues:

```
Probably at least one of the constraints in the following list is one you don't want. Try this: (1) look at each constraint and try to figure out which you don't expect; (2) find the code that added the unwanted constraint or constraints and fix it. (Note: If you're seeing NSAutoresizingMaskLayoutConstraints that you don't understand, refer to the documentation for the UIView property translatesAutoresizingMaskIntoConstraints) 
(
    "<NSLayoutConstraint: 0x7fa4aac7b6d0> H:|-(20)-[UIView:0x7fa4aab42410] ((null), (null))",
    "<NSLayoutConstraint: 0x7fa4aac7b720> H:[UIView:0x7fa4aab42410]-(20)-| ((null), (null))",
    "<NSLayoutConstraint: 0x7fa4aac80dd0> (null) ((null), (null))",
    "<NSLayoutConstraint: 0x7fa4aac81230> H:|-(0)-[Artsy.LotStandingsLotListView:0x7fa4aac752d0] ((null), (null))",
    "<NSLayoutConstraint: 0x7fa4aac813f0> (null) ((null), (null))",
    "<NSLayoutConstraint: 0x7fa4aac82300> (null) ((null), (null))",
    "<NSLayoutConstraint: 0x7fa4aae32c80> H:|-(0)-[ORTagBasedAutoStackView:0x7fa4aac460a0] ((null), (null))",
    "<NSLayoutConstraint: 0x7fa4aae37f50> (null) ((null), (null))",
    "<NSLayoutConstraint: 0x7fa4aad56b90> H:[UICollectionReusableView:0x7fa4aaca4bf0(0)] ((null), (null))"
)

Will attempt to recover by breaking constraint 
<NSLayoutConstraint: 0x7fa4aac7b720> H:[UIView:0x7fa4aab42410]-(20)-| ((null), (null))
```

Could be related to a layout issue Katarina showed in her feedback. Recording here so I'll remember to circle back on it.

**Comment by ashfurrow** [comment]
Design feedback done:

screen shot 2017-03-07 at 11 40 35 am

Now to fix those bugs!

**Comment by ashfurrow** [comment]
Okay, this is ready for review. @orta is away, going to get some input from @mennenia.

File: Artsy/View_Controllers/Embedded/Generics/AREmbeddedModelsViewController.m [hunk 9]
```diff
@@ -350,8 +350,7 @@ - (UICollectionReusableView *)collectionView:(UICollectionView *)collectionView
         UICollectionReusableView *view = [collectionView dequeueReusableSupplementaryViewOfKind:kind withReuseIdentifier:kind forIndexPath:indexPath];
         if (view.subviews.count == 0) {
             [view addSubview:self.stickyHeaderView];
-            [self.stickyHeaderView alignTopEdgeWithView:view predicate:@"0"];
-            [self.stickyHeaderView alignLeading:@"0" trailing:@"0" toView:view];
+            [self.stickyHeaderView alignToView:view];
```
Comment (mennenia) [inline Artsy/View_Controllers/Embedded/Generics/AREmbeddedModelsViewController.m]: This is great! I didn't know there was `alignToView` lol

File: Podfile.lock [hunk 10]
```diff
@@ -423,4 +423,4 @@ SPEC CHECKSUMS:
 
 PODFILE CHECKSUM: a4c1a229f7ad09b360fa6ba10dc0917a95ff8239
 
-COCOAPODS: 1.2.0
+COCOAPODS: 1.1.1
```
Comment (mennenia) [inline Podfile.lock]: Was this an accidental or purposeful change? :o

**Comment by mennenia** [comment]
@ashfurrow I do have one Q: in one of the [snapshots](https://github.com/artsy/eigen/pull/2162/files#diff-e973c1d26251194f854a4ca0240ea9f5) there's a bit where "highest bidder" is cut off:

screen shot 2017-03-09 at 16 59 57

Is that supposed to be like that? I know sometimes the snapshot tests are to verify other layout elements, so it could of course be on purpose! :)

**Comment by ashfurrow** [comment]
@mennenia yikes, yeah that's exactly what the snapshot is (using iPad UI in iPhone dimensions). I'll take a look 👍

File: Podfile.lock (hunk 10 above)
Comment (ashfurrow) [inline Podfile.lock]: On purpose I think, the Gemfile still has the old CocoaPods version ¯\\_(ツ)_/¯

**Comment by ashfurrow** [comment]
Okay, it looks like the autolayout bugs are unrelated to this PR, going to merge this and follow up with the errors in a separate PR.

**Pull request merged**